adapters/sovrn/sovrn.go

```diff
@@ -0,0 +1,274 @@
+
+	if sResp.StatusCode != 200 {
+		return nil, fmt.Errorf("HTTP status %d; body: %s", sResp.StatusCode, responseBody)
+	}
+
+	if req.IsDebug {
+		debug.ResponseBody = responseBody
+	}
+
+	var bidResp openrtb.BidResponse
+	err = json.Unmarshal(body, &bidResp)
+	if err != nil {
+		return nil, err
+	}
+
+	bids := make(pbs.PBSBidSlice, 0)
+
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			bidID := bidder.LookupBidID(bid.ImpID)
+			if bidID == "" {
+				return nil, fmt.Errorf("Unknown ad unit code '%s'", bid.ImpID)
+			}
+
+			adm, _ := url.QueryUnescape(bid.AdM)
+			pbid := pbs.PBSBid{
+				BidID:       bidID,
+				AdUnitCode:  bid.ImpID,
+				BidderCode:  bidder.BidderCode,
+				Price:       bid.Price,
+				Adm:         adm,
+				Creative_id: bid.CrID,
+				Width:       bid.W,
+				Height:      bid.H,
+				DealId:      bid.DealID,
+				NURL:        bid.NURL,
+			}
+			bids = append(bids, &pbid)
+		}
+	}
+
+	sort.Sort(bids)
+	return bids, nil
+}
+
+func (s *SovrnAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+
+	for i := 0; i < len(request.Imp); i++ {
+		_, err := preprocess(&request.Imp[i])
+		if err != nil {
+			errs = append(errs, err)
+			request.Imp = append(request.Imp[:i], request.Imp[i+1:]...)
+			i--
+		}
+	}
+
+	// If all the requests were malformed, don't bother making a server call with no impressions.
+	if len(request.Imp) == 0 {
+		return nil, errs
+	}
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	headers.Add("User-Agent", request.Device.UA)
+	headers.Add("X-Forwarded-For", request.Device.IP)
+	headers.Add("Accept-Language", request.Device.Language)
+	headers.Add("DNT", strconv.Itoa(int(request.Device.DNT)))
+
+	userID := strings.TrimSpace(request.User.BuyerUID)
+	if len(userID) > 0 {
+		headers.Add("Cookie", fmt.Sprintf("%s=%s", "ljt_reader", userID))
+	}
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     s.URI,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+func (s *SovrnAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) ([]*adapters.TypedBid, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bids := make([]*adapters.TypedBid, 0, 5)
+
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			bids = append(bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: getMediaTypeForImp(bid.ImpID, internalRequest.Imp),
```

Comment:
I'm a little confused... you're only accepting `Banner` imps in the `preprocess` method. Shouldn't you just hardcode `Banner` here?

Reply:
This mimics the look of other adapters, as well as gives us a little bit of future proofing, in case we start supporting other types. I can hard-code though if you think it will lessen confusion

Reply:
yeah... I think the code should be written as simply as possible, given the adapter's current capabilities.

The code here ("Assume the bid is video, as long as it could be") is really only valid for Appnexus... and even then, not valid anymore because Appnexus updated their bid responses so we could make it less hacky (see #325).

Definitely wouldn't copy it unless it's intentional & your server also bids with the same logic.